Chap_API_Sets_Groups.tex

```diff
@@ -4,37 +4,125 @@
+corresponding \acp{API}.
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 \section{Process Sets}
 \label{chap:api_sets_groups:sets}
 
-A \ac{PMIx} \emph{Process Set} is a user-provided label associated with a given set of application processes. Processes can belong to multiple process \emph{sets} at a time. Definition of a \ac{PMIx} process set typically occurs at time of application execution - e.g., on a command line:
+A \ac{PMIx} \emph{Process Set} is a user-provided or host environment assigned
+label associated with a given set of application processes. Processes can
+belong to multiple process \emph{sets} at a time. Definition of a \ac{PMIx}
+process set typically occurs at time of application execution - e.g., on a
+command line:
 
 \cspecificstart
 \begin{codepar}
 \$ prun -n 4 --pset ocean myoceanapp : -n 3 --pset ice myiceapp
 \end{codepar}
 \cspecificend
 
-In this example, the processes in the first application will be labeled with a \refattr{PMIX_PSET_NAME} attribute of \emph{ocean} while those in the second application will be labeled with an \emph{ice} value. During the execution, application processes could lookup the process set attribute for any other process using \refapi{PMIx_Get}. Alternatively, other executing applications could utilize the \refapi{PMIx_Query_info_nb} \ac{API} to obtain the number of declared process sets in the system, a list of their names, and other information about them. In other words, the \emph{process set} identifier provides a label by which an application can derive information about a process and its application - it does \emph{not}, however, confer any operational function.
+In this example, the processes in the first application will be labeled with a \refattr{PMIX_PSET_NAMES} attribute of \emph{ocean} while those in the second application will be labeled with an \emph{ice} value. During the execution, application processes could lookup the process set attribute for any other process using \refapi{PMIx_Get}. Alternatively, other executing applications could utilize the \refapi{PMIx_Query_info_nb} \ac{API} to obtain the number of declared process sets in the system, a list of their names, and other information about them. In other words, the \emph{process set} identifier provides a label by which an application can derive information about a process and its application - it does \emph{not}, however, confer any operational function.
 
-Thus, process \emph{sets} differ from process \emph{groups} in several key ways:
+Host environments can create or delete process sets at any time, as well as
+add and delete members from existing process sets, through the
+\refapi{PMIx_server_define_process_set} and
+\refapi{PMIx_server_delete_process_set} \acp{API}. \ac{PMIx} servers shall
+notify all local clients of process set operations via the
+\refconst{PMIX_PROCESS_SET_DEFINE} or \refconst{PMIX_PROCESS_SET_DELETE}
+events.
```

Comment:
What happens to groups that are based on a given set if the set membership changes at runtime?  Are the group and set membership kept in sync automatically by PMIx?  Or is it the job of users of the groups to subscribe to set membership changes and do bookkeeping themselves? (I may be confusing the interfaces of MPI Sessions and PMIx so my apologies if this question doesn't make sense).

Reply:
What may help to clarify my confusion above is a better understanding of how a group is created from a set.  Is the idea that if I want to make a group based on the processes from the set `topology:/rack-1` (as a random example), that I would first query set membership with the `PMIX_QUERY_GROUP_MEMBERSHIP` key to get the process IDs in the set and then do a `PMIx_Group_construct` with that listing of process IDs?  Or is it possible to pass the set name `topology:/rack-1` directly into `PMIx_Group_construct`?

Reply:
I think the basic misunderstanding here is that groups are _not_ created from sets - groups are created from an array of proc IDs. So change in membership of one (whether a group or a set) has no impact on membership of the other, even if the user gives the group the same name as the host environment assigned to some collection of procs and both have the same procs as members.

>  I would first query set membership with the PMIX_QUERY_GROUP_MEMBERSHIP key to get the process IDs in the set and then do a PMIx_Group_construct with that listing of process IDs

Well, it would actually be `PMIX_QUERY_PSET_MEMBERSHIP`, but your method is correct. However, as stated above, that would not imply connection between the set and group.

> Or is it possible to pass the set name `topology:/rack-1` directly into `PMIx_Group_construct`?

No - remember, process sets have no operational definition, meaning that you cannot use the set name in any PMIx operation. Only process groups have operational meaning. However, you could pass that string into `PMIx_Group_construct` to name your new group `topology:/rack-1` if you wanted to "tag" the group as being based on that set - it just wouldn't mean anything to PMIx.

Does that help? Is there some change in wording required to avoid the misunderstanding? Just wondering what took you down this path.

Reply:
Added a few words to clarify that there is no linkage between process sets and groups of the same name, even if they are formed from the same members, so changes in one aren't automatically reflected in the other.

Reply:
> Does that help?

Yes. That does.  Thanks for the clarifications.

> Just wondering what took you down this path.

Turns out I was getting MPI sets/groups semantics mixed up with PMIx's sets/groups. Thanks @hppritcha for reviewing the MPI sessions semantics during the functionality WG meeting.